### PR TITLE
push blocks to stream each block

### LINKED ISSUE #604: [Sequencer] Update datastream block-by-block if no external executors set
Currently we update the stream batch-by-batch which causes a strange pattern of RPC updates.

To make it behave similarly like the legacy node, let’s update it block by block.

Only if there are no external executors set.

### PR DESCRIPTION
**Pull request by V-Staykov**

### CODE REVIEW

**Comment by mandrigin** [comment]
converted to draft until we solve the sequencer restart problem

File: core/vm/zk_batch_counters.go [hunk 1]
```diff
@@ -208,3 +217,42 @@ func (bcc *BatchCounterCollector) CombineCollectors(verifyMerkleProof bool) (Cou
 
 	return combined, nil
 }
+
+// CombineCollectors takes the batch level data from all transactions and combines these counters with each transactions'
+// rlp level counters and execution level counters
+func (bcc *BatchCounterCollector) CombineCollectorsNoChanges(externalUsedCounters *Counters, verifyMerkleProof bool) Counters {
```
Comment (hexoscott) [inline core/vm/zk_batch_counters.go]: Could you update the comment to represent the distinction between `CombineCollectors`

File: zk/stages/stage_sequence_execute.go [hunk 2]
```diff
@@ -475,31 +509,70 @@ func SpawnSequencingStage(
...
+			if err = tx.Commit(); err != nil {
+				return err
+			}
+			tx, err = cfg.db.BeginRw(ctx)
+			if err != nil {
+				return err
+			}
+			// TODO: This creates stacked up deferrals
+			defer tx.Rollback()
+			sdb.SetNewTx(tx)
```
Comment (hexoscott) [inline zk/stages/stage_sequence_execute.go]: would this be better positioned after the write to the datastream?  Just thinking aloud, if the write to the stream fails we will have already committed the block details and counters to the DB.  If we do it after the write to the stream then everything gets commited as one, any issue committing the DB stuff will cause an unwind in the stream anyway which we're already handling.

File: cmd/rpcdaemon/commands/zkevm_counters.go [hunk 3]
```diff
@@ -149,7 +149,7 @@ func (zkapi *ZkEvmAPIImpl) EstimateCounters(ctx context.Context, rpcTx *zkevmRPC
 	txCounters := vm.NewTransactionCounter(tx, int(smtDepth), uint16(forkId), zkapi.config.Zk.VirtualCountersSmtReduction, false)
 	batchCounters := vm.NewBatchCounterCollector(int(smtDepth), uint16(forkId), zkapi.config.Zk.VirtualCountersSmtReduction, false)
 
-	_, err = batchCounters.AddNewTransactionCounters(txCounters)
+	_, err = batchCounters.AddNewTransactionCounters(nil, txCounters)
```
Comment (hexoscott) [inline cmd/rpcdaemon/commands/zkevm_counters.go]: There are quite a few methods that now take the existing counters as a parameter.  Would it be possible to add the existing counters to the constructor of the batchCounters itself to save affecting the methods?

File: zk/stages/stage_sequence_execute.go [hunk 4]
```diff
@@ -146,7 +174,7 @@ func SpawnSequencingStage(
 		limboHeaderTimestamp = math.MaxUint64
 	}
 
-	if l1Recovery {
+	if l1Recovery && isLastBatchFinished {
```
Comment (hexoscott) [inline zk/stages/stage_sequence_execute.go]: will this cause the recovery to stop after the first batch has been processed during the recovery?

**Review by hexoscott — COMMENTED** [review]
looks good!  Some small comments.

File: zk/stages/stage_sequence_execute.go (hunk 4 above)
Comment (V-Staykov) [reply zk/stages/stage_sequence_execute.go]: I expect this to finish last started batch before going into recovery. Now that I think of it there is a case where te batch can't be finished and hangs, so maybe we should do this some other way.

Comment (hexoscott) [reply zk/stages/stage_sequence_execute.go]: Thinking about it again this morning, I think we can just remove the `isLastBatchFinished` part.  L1 recovery shouldn't get into a position where we can't finish the batch.

Comment (V-Staykov) [reply zk/stages/stage_sequence_execute.go]: Okay, going to remove it

**Review by hexoscott — DISMISSED** [review]

File: turbo/stages/stageloop.go [hunk 5]
```diff
@@ -151,6 +151,8 @@ func StageLoopStep(ctx context.Context, chainConfig *chain.Config, db kv.RwDB, s
 	}
 	canRunCycleInOneTransaction := !initialCycle
 
+	canRunCycleInOneTransaction = false // we need to commit when sequencer each run
```
Comment (hexoscott) [inline turbo/stages/stageloop.go]: does this need to be surrounded by a check for `isSequencer` in this case?

**Review by revitteth — DISMISSED** [review]

File: act
Comment (hexoscott) [inline act:1]: how come we've added a submodule?

**Review by hexoscott — APPROVED** [review]

**Pull request merged**